src/crimson/os/seastore/lba_mapping.h

```diff
@@ -4,69 +4,208 @@
 #pragma once
 
 #include "crimson/os/seastore/cached_extent.h"
-#include "crimson/os/seastore/btree/btree_range_pin.h"
+#include "crimson/os/seastore/btree/btree_types.h"
 #include "crimson/os/seastore/lba_manager/btree/lba_btree_node.h"
 #include "crimson/os/seastore/logical_child_node.h"
 
 namespace crimson::os::seastore {
 
-class LBAMapping;
-using LBAMappingRef = std::unique_ptr<LBAMapping>;
+namespace lba_manager::btree {
+class BtreeLBAManager;
+}
 
-class LogicalCachedExtent;
-
-class LBAMapping : public BtreeNodeMapping<laddr_t, paddr_t> {
+class LBAMapping {
 public:
-  LBAMapping(op_context_t<laddr_t> ctx)
-    : BtreeNodeMapping<laddr_t, paddr_t>(ctx) {}
-  template <typename... T>
-  LBAMapping(T&&... t)
-    : BtreeNodeMapping<laddr_t, paddr_t>(std::forward<T>(t)...)
-  {}
-
-  // An lba pin may be indirect, see comments in lba_manager/btree/btree_lba_manager.h
-  virtual bool is_indirect() const = 0;
-  virtual laddr_t get_intermediate_key() const = 0;
-  virtual laddr_t get_intermediate_base() const = 0;
-  virtual extent_len_t get_intermediate_length() const = 0;
-  // The start offset of the pin, must be 0 if the pin is not indirect
-  virtual extent_len_t get_intermediate_offset() const = 0;
-
-  virtual get_child_ret_t<lba_manager::btree::LBALeafNode, LogicalChildNode>
-  get_logical_extent(Transaction &t) = 0;
-
-  void link_child(LogicalChildNode *c) {
-    ceph_assert(child_pos);
-    child_pos->link_child(c);
-  }
-  virtual LBAMappingRef refresh_with_pending_parent() = 0;
+  LBAMapping() = default;
+  LBAMapping(LBACursorRef direct, LBACursorRef indirect)
+    : direct_cursor(std::move(direct)),
+      indirect_cursor(std::move(indirect))
+  {
+    assert(!is_null());
+    assert(!direct_cursor || !direct_cursor->is_indirect());
+    assert(!indirect_cursor || indirect_cursor->is_indirect());
+  }
+
+  static LBAMapping create_direct(LBACursorRef iter) {
+    return LBAMapping(std::move(iter), nullptr);
+  }
+
+  static LBAMapping create_indirect(LBACursorRef iter) {
+    return LBAMapping(nullptr, std::move(iter));
+  }
+
+  LBAMapping(const LBAMapping &) = delete;
+  LBAMapping(LBAMapping &&) = default;
+  LBAMapping &operator=(const LBAMapping &) = delete;
+  LBAMapping &operator=(LBAMapping &&) = default;
+  ~LBAMapping() = default;
+
+  bool is_null() const {
+    return !direct_cursor && !indirect_cursor;
+  }
+
+  bool is_direct() const {
+    assert(!is_null());
+    return direct_cursor && !indirect_cursor;
+  }
+
+  bool is_linked_direct() const {
+    assert(!is_null());
+    return (bool)direct_cursor;
+  }
+
+  bool is_indirect() const {
+    assert(!is_null());
+    if (indirect_cursor) {
+      // XXX: direct_cursor could be null in the future,
+      // but we treat it as error for now.
```

Comment:
See other comments, seems to me any future use case that requires a valid mapping to have a null direct_cursor is more appropriate to use (indirect) cursors instead. What do you think ?